design-proposals/garbage-collect-completed-dvs.md

```diff
@@ -0,0 +1,59 @@
+# Overview
+DataVolumes are a useful construct for provisioning VM disks but once the PVC is provisioned and populated the DataVolume provides no value. This design presents an approach for garbage collecting completed DataVolumes. 
+
+## Motivation
+In the field, customers are confused by DVs after the fact. For example, when someone deletes a PVC that is owned by a DV they are confused when the PVC is re-created.
+
+## Goals
+DataVolumes will be garbage collected after their PVC is provisioned and populated.
+
+## Non Goals
+(limitations to the scope of the design)
+
+## Definition of Users
+(who is this feature set intended for)
+
+## User Stories
+* As a VM owner I use a standalone DataVolume to prepare a VM disk PVC for use by my VM. Once the DataVolume is complete, it has no use and can be confusing, so I want the system to garbage collect it. The PVC should still maintain an independent lifecycle from the VM.
+* As a VM owner I add an entry to the dataVolumeTemplates section of my VM spec to have a disk prepared for my VM. Once the DataVolume is complete, it has no use and can be confusing, so I want the system to garbage collect it. I want kubevirt to understand that it doesn't need to create a new DataVolume when the underlying PVC exists. The PVC should still share the same lifecycle as the VM.
+* As a VM owner I want to restore my backed up VM without needing to recreate the DataVolume.
+
+## Repos
+* **CDI**: Controllers (DataVolume/Import/Upload/Clone/DataImportCron/DataSource) will need adaptations as they currently assume existance of DV-PVC pair
+* **KubeVirt**: VM controller adaptations
+
+# Design
+
+## PVC Ownership
+The DataVolume controller creates a PVC owned by its DV. When the PVC is populated (DV is succeeded):
+* PVC ownership by the DV will be removed
+* If DV has ownership (e.g. by VM, when created via DataVolumeTemplates), it will be transferred to its PVC
+* Any DV information that might be useful after it is garbage collected, should be annotated on the PVC (TBD)
+
+If the PVC population failed, the DV should remain the owner of the PVC.
+
+## Garbage Collection
+When DataVolume does not own a PVC, it can be deleted immediately, but we may want to keep it for a while:
+* Add CDI config a default time (minimal, zero if not set) for DV garbage collection after completion (TBD: why needed?)
+  `dvTTLSecondsAfterCompletion *int32`
+* Add DV annotation to prevent its garbage collection
+  `"cdi.kubevirt.io/storage.dv.retainAfterCompletion": "true"`
```

Comment:
If we make the ttl setting global on the CDI config (as a means of opting into GC) then i don't know if we have to have a per DV ttl setting.

The annotation here is probably good though. It gives us an escape hatch for disabling GC on data volumes in the field if we discover an issue that requires the data volume to stick around.